container.go

```diff
@@ -109,90 +97,118 @@ func (c Container) invoke(function interface{}) (interface{}, error) {
 	return values[0].Interface(), nil
 }
 
+// make resolves the binding and returns the concrete.
+// Search up any parent container scopes if the binding is not found in current scope.
+func (c *Container) make(t reflect.Type, name string) (interface{}, error) {
+	current := c
+	var binding *binding
+
+	for {
+		if found, exist := current.bindings[t][name]; exist {
+			binding = found
+			break
+		}
+
+		if current.parent == nil {
+			break
+		} else {
+			current = current.parent
+		}
+	}
+
+	if binding == nil {
+		return nil, errors.New("container: no binding found for: " + t.String())
+	}
+
+	return binding.make(current)
+}
+
 // arguments returns the list of resolved arguments for a function.
-func (c Container) arguments(function interface{}) ([]reflect.Value, error) {
+func (c *Container) arguments(function interface{}) ([]reflect.Value, error) {
 	reflectedFunction := reflect.TypeOf(function)
 	argumentsCount := reflectedFunction.NumIn()
 	arguments := make([]reflect.Value, argumentsCount)
 
 	for i := 0; i < argumentsCount; i++ {
 		abstraction := reflectedFunction.In(i)
-		if concrete, exist := c[abstraction][""]; exist {
-			instance, err := concrete.make(c)
-			if err != nil {
-				return nil, err
-			}
+
+		if instance, err := c.make(abstraction, ""); err == nil {
 			arguments[i] = reflect.ValueOf(instance)
 		} else {
-			return nil, errors.New("container: no concrete found for: " + abstraction.String())
+			return nil, fmt.Errorf("container: encountered error while making instance for: %s. Error encountered: %w", abstraction.String(), err)
 		}
 	}
 
 	return arguments, nil
 }
 
 // Reset deletes all the existing bindings and empties the container.
-func (c Container) Reset() {
-	for k := range c {
-		delete(c, k)
+func (c *Container) Reset() {
+	for k := range c.bindings {
```

Comment:
given that `make` will walk up the parent chain should `Reset()` do something with `parent` here?

Reply:
We can make reset go up the stack and reset all the bindings but that may not achieve desired result.  For resolving bindings it does make sense that it should look at each parent and determine if their is a binding available I'm just not sure if I call reset on a child scope that that would be the intended behavior vs only clearing out the bindings on the current scope.

This is primarily used for testing and I don't think we have any scenario in `azd` where we call reset.

Reply:
I think you're right that navigating up the parent's stack and doing a reset is likely more dangerous than not.  Good to know this is just used in Testing, it seems like it is very hard for us to provide a reasonable semantics for this now that we have this parent chain.  I'm happy with keeping what you had here.  Thanks for the explanation.